### ISSUE TITLE
Contextual Contrasting Loss Function

### ISSUE THREAD

**Issue by antonia42** [emadeldeen24/TS-TCC#18]
Hi there,

I was taking a closer look in the implementation of the contextual contrasting loss and and I am having trouble understanding how the positive samples are being treated differently from the negative ones, and specifically how this corresponds to the Eq. 5 of your paper. Could it be that `labels` should have some elements equal to one and not all of them being zeros?

https://github.com/emadeldeen24/TS-TCC/blob/ebfdbab03dc03842b0c18ec16d64fbd923bd5779/models/loss.py#L61-L62

Thanks!

**Comment by emadeldeen24** [comment]
In the contextual contrasting loss, after calculating the similarity matrix for elements in zis and zjs, we extract the positive and negative samples from the similarity matrix. 
The positive samples are the diagonal elements of the upper-left and lower-right quadrants, corresponding to elements that come from the same representation. These positive samples are extracted twice: once with an offset of `batch_size` to get the positive samples from the upper-left quadrant, and once with an offset of `-batch_size` to get the positive samples from the lower-right quadrant. The positive samples are then concatenated and reshaped into a column vector called positives.

The negative samples are obtained by masking the similarity matrix using the `mask_samples_from_same_repr` mask, which has a value of True for elements that should be considered negative samples. 
The logits are constructed by concatenating positives and negatives along the second dimension (columns). 

The labels for the logits are created as a tensor of zeros with a shape of (2 * batch_size). Since the positive samples are in the first column of the logits, the labels are all zeros, indicating that the first column of logits corresponds to the positive samples.

In summary, the positive samples are treated differently by extracting them separately from the similarity matrix and concatenating them as the first column of the logits. The negative samples are obtained by masking the similarity matrix, and they occupy the remaining columns of the logits. The labels are set to zeros for all samples, indicating that the first column corresponds to the positive samples.
